Split middlewares

yogsototh opened the pull request:
> Related https://github.com/advthreat/iroh/pull/5749

Code refacto to split the single main `wrap-jwt` middleware in two `wrap-authorization` and `wrap-authentication`.

Removed a lot of unnecessary injection and coupling.

- Removed the dependency from `clj-momo`. Now we only use `clojure.java-time` for tests.
- Added schemas.
- Prevent log function injection (and some others) by preventing exceptions.

yogsototh committed:
Split middlewares

yogsototh committed:
toward removing clj-momo deps

yogsototh committed:
java-time

yogsototh committed:
wip

yogsototh committed:
monadic structre, prevent exceptions

yogsototh committed:
schemas + abstractions

yogsototh committed:
Cleanup exceptions

yogsototh committed:
Get rid of logs deps

yogsototh committed:
Update README

yogsototh committed:
Added schemas

yogsototh committed:
cleanup and docstrings

yogsototh committed:
Use monadic style macro

frenchy64 commented on src/ring_jwt_middleware/core.clj:
```diff
@@ -233,68 +276,85 @@
+                                    (->pure :ok))
+                                  (catch Exception e
+                                    (->err :jwt-revocation-fn-exception
+                                           "is-revoked-fn thrown an exception"
```
threw

frenchy64 commented on src/ring_jwt_middleware/core.clj:
```diff
@@ -74,84 +100,100 @@
-         (str/join " "))))
+         (string/join " "))))
+
+(s/defn current-epoch! :- s/Num
```
Did you consider threading an optional `current-epoch!` argument to the middleware to override using lexical scope? Perhaps we can provide an alternative stubbing API from `with-redefs`.

frenchy64 commented on src/ring_jwt_middleware/core.clj:
```diff
@@ -74,84 +100,100 @@
+    jwt-max-lifetime-in-sec :- s/Num
+    jwt-check-fn]
+   (let-either [_ (check-jwt-expiry jwt (or jwt-max-lifetime-in-sec default-jwt-lifetime-in-sec))]
+     (if (fn? jwt-check-fn)
```
Could we loosen the spec to allow vars and multimethods and such? I see there's an error log if it's not a fn so it doesn't seem urgent.

frenchy64 commented on src/ring_jwt_middleware/core.clj:
```diff
@@ -233,68 +276,85 @@
+          (handler (into request (<-result authentication-result))))))))
+
+(defn mk-wrap-authorization
+  "A function building a middleware taking care of the authorization logic.
```
Can you help me understand what this function has to do with authorization? All I see are error handlers and special case handling.

yogsototh replied:
The authentication will just take care of adding a few fields to the ring request.

The authorization is responsible to make a call fail or not if the request contains an authentication error or not.

Generally the authorization logic should not be the concern of this lib but external.
Still, there is a best practice we could enforce by making the distinction between request without any authentication, and request with failed authentication attempts.

For the second case, the request should fail, and this is enforced by using a function without any `handler` as an argument.

frenchy64 reviewed (approved):
LGTM

msprunck commented on src/ring_jwt_middleware/core.clj:
```diff
@@ -233,68 +276,85 @@
-            (no-jwt-fn request)))))))
+        (fn [{{:keys [error] :as jwt-error} :jwt-error
+              :as request}]
+          (if jwt-error
```
Instead of checking the `jwt-error` I would check the presence of the `identity` or the `jwt` key to know if the request is authenticated. Not having a `jwt-error` doesn't mean that the request is authenticated.

You could extract an `authenticated?` function to better expose the logic in this middleware. I saw something similar in [buddy-auth](https://github.com/funcool/buddy-auth/blob/f8ff8693ed2b328e913a77806e20ebd8142a1dbe/src/buddy/auth.clj#L27)

msprunck commented on src/ring_jwt_middleware/core.clj:
```diff
@@ -233,68 +276,85 @@
+    (fn [handler]
+      (fn [request]
+        (let [authentication-result
+              (let-either [raw-jwt (get-jwt request)
```
Much better than the nested `if-let`s

msprunck reviewed (approved):
LGTM!

agzam commented on .gitignore:
```diff
@@ -11,3 +11,4 @@ pom.xml.asc
 .hg/
 .projectile
 .*.swp
+.clj-kondo/.cache/
```
While you're at it, could you please also add `.lsp/.cache/` as well?

agzam commented on src/ring_jwt_middleware/result.clj:
```diff
@@ -0,0 +1,93 @@
+    (:result result)))
+
+(defmacro let-either
+  "The let-either macro can be used to handle cascading results that
+  can depend on preceding values.
+  If one of the function fail, we return the failed result.
+  If all functions are successful we return the content of the
+  body."
```
```suggestion
  "To be used to handle cascading results that may depend on preceding values.
  If one of the functions fails, it returns the failed result.
  If all functions are successful, it returns the content of the body."
```

agzam reviewed (approved):
This is beautiful work. It would take me far too long to learn the context and design decisions that went into this and review it more thoroughly. But I see that @frenchy64  and @msprunck  are handling it from multiple angles, so there's nothing left for me to pick up here :). At this point, I can only appreciate the beauty and approve this work. Good job, Yann!

agzam commented on src/ring_jwt_middleware/result.clj:
```diff
@@ -0,0 +1,93 @@
+(ns ring-jwt-middleware.result
+  "This ns provide a set of helpers to handle an abstraction similar to Either in Haskell
+
+  The main goal is to provide a mechanism similar to the exceptions but pure without Java Exceptions.
+
+  A function that return a `Result` means the function returned either a successful result or an error with a
+  common error structure (`JwtError`).
```
```suggestion
  "This ns provides a set of helpers to handle an abstraction similar to Either in Haskell.

  The main goal is to provide a mechanism similar to the exceptions, albeit pure - without Java Exceptions.

  When a function returns a `Result` - it either contains a result of a successful outcome or an error with a
  common error structure (`JwtError`).
```

agzam commented on src/ring_jwt_middleware/result.clj:
````diff
@@ -0,0 +1,93 @@
+    ,,,)
+  ```
+
+  if `fn-returning-a-result-1` return an error then we will not execute the rest of the let-either.
````
```suggestion
  if `fn-returning-a-result-1` returns an error then we will not execute the rest of the let-either.
```

agzam commented on README.org:
```diff
@@ -10,57 +10,91 @@ currently, only RS256 is supported.
 - JWT lifetime & Expiration support
-- custom additional validation through a user provided fn
-- custom revokation check through a user provided fn
+- custom additional validation through a user provided functions
+- custom revocation check through a user provided functions
```
```suggestion
- custom additional validation through user-provided functions
- custom revocation check through user-provided functions
```

agzam commented on README.org:
```diff
@@ -10,57 +10,91 @@ currently, only RS256 is supported.
+- ~jwt-error~ => will contain an error object if the something went wrong with the JWT
+
+The ~wrap-authentication~ will not take any decision about authorization access.
+This lib also provide another helper to build another middleware handling
```
```suggestion
This lib also provides another helper to build another middleware handling
```

agzam commented on src/ring_jwt_middleware/result.clj:
```diff
@@ -0,0 +1,93 @@
+   (s/optional-key :jwt-error) JwtError})
+
+(s/defschema Result
+  ;; A result is similar to the Either in Haskell
+  ;; It represent either a value or an error
```
It looks like s/defschema takes docstring. Maybe you meant this comment to be a docstring?

yogsototh committed:
wip

yogsototh committed:
wip

yogsototh committed:
wip

yogsototh committed:
Another round of refacto

- split schemas
- split config related
- code architecture purity
- test improvements

yogsototh committed:
doc

yogsototh committed:
Update src/ring_jwt_middleware/result.clj

Co-authored-by: Ag Ibragimov <agzam.ibragimov@gmail.com>

yogsototh committed:
Apply suggestions from code review

Co-authored-by: Ag Ibragimov <agzam.ibragimov@gmail.com>

yogsototh committed:
More suggestions applied

yogsototh committed:
comment to docstring

yogsototh committed:
Improved README
